File: src/fx_spec/fx_spec.ml
open FastCheck
open Fx
open Vitest

let () =
  test "Ok" (fun () ->
      (fun input ->
        match Ok input |. run ~deps:() with
        | Error _ -> unreachable ()
        | Ok value -> expect value |. toBe input)
      |> property1 (anything ())
      |> assert_);

  test "Error" (fun () ->
      (fun input ->
        match Error input |. run ~deps:() with
        | Error error -> expect error |. toBe input
        | Ok _ -> unreachable ())
      |> property1 (anything ())
      |> assert_);

  testAsync "async" (fun () ->
      let open Promise.Monad_syntax in
      let* () =
        (fun input ->
          let* result =
            Async (fun n -> Ok n |> Js.Promise.resolve) |. runAsync ~deps:input
          in
          (match result with
          | Error _ -> unreachable ()
          | Ok value -> expect value |. toBe input);
          Js.Promise.resolve ())
        |> asyncProperty1 (anything ())
        |> asyncAssert
      in
      let* () =
        (fun input ->
          let* result =
            Async (fun n -> Error n |> Js.Promise.resolve)
            |. runAsync ~deps:input
          in
          (match result with
          | Error error -> expect error |. toBe input
          | Ok _ -> unreachable ());
          Js.Promise.resolve ())
        |> asyncProperty1 (anything ())
        |> asyncAssert
      in
      Js.Promise.resolve ());

  testAsync "bind" (fun () ->
      let open Promise.Monad_syntax in
      let () =
        (fun input ->
          match
            Ok input
            |. bind (fun n -> Ok (n |> Js.Float.toString))
            |. run ~deps:()
          with
          | Error _ -> unreachable ()
          | Ok value -> expect value |. toBe (Js.String.make input))
        |> property1 (double ())
        |> assert_
      in
      let () =
        (fun input ->
          let f = fn1 () in
          (match Error input |. bind f |. run ~deps:() with
          | Error error -> expect error |. toBe input
          | Ok _ -> unreachable ());
          expect f |. not |. toBeCalled)
        |> property1 (string ())
        |> assert_
      in
      let* () =
        (fun input ->
          let* result =
            Async (fun () -> Ok input |> Js.Promise.resolve)
            |. bind (fun n -> Ok (n |> Js.Float.toString))
            |. runAsync ~deps:()
          in
          (match result with
          | Error _ -> unreachable ()
          | Ok value -> expect value |. toBe (Js.String.make input));
          Js.Promise.resolve ())
        |> asyncProperty1 (double ())
        |> asyncAssert
      in
      let* () =
        (fun input ->
          let f = fn1 () in
          let* result =
            Async (fun () -> Error input |> Js.Promise.resolve)
            |. bind f |. runAsync ~deps:()
          in
          (match result with
          | Error error -> expect error |. toBe input
          | Ok _ -> unreachable ());
          expect f |. not |. toBeCalled;
          Js.Promise.resolve ())
        |> asyncProperty1 (string ())
        |> asyncAssert
      in
      Js.Promise.resolve ());

  testAsync "map" (fun () ->
      let open Promise.Monad_syntax in
      let () =
        (fun input ->
          match
            Ok input |. map (fun n -> n |> Js.String.make) |. run ~deps:()
          with
          | Error _ -> unreachable ()
          | Ok value -> expect value |. toBe (input |. Js.Float.toString))
        |> property1 (double ())
        |> assert_
      in
      let () =
        (fun input ->
          let f = fn1 () in
          let result = Error input |. bind f |. run ~deps:() in
          (match result with
          | Error error -> expect error |. toBe input
          | Ok _ -> unreachable ());
          expect f |. not |. toBeCalled)
        |> property1 (string ())
        |> assert_
      in
      let* () =
        (fun input ->
          let* result =
            Ok input
            |. bind (fun n -> Ok (n |> Js.Float.toString))
            |. runAsync ~deps:()
          in
          (match result with
          | Error _ -> unreachable ()
          | Ok value -> expect value |. toBe (input |> Js.Float.toString));
          Js.Promise.resolve ())
        |> asyncProperty1 (double ())
        |> asyncAssert
      in
      let* () =
        (fun input ->
          let f = fn1 () in
          let* result =
            Async (fun () -> Error input |> Js.Promise.resolve)
            |. bind f |. runAsync ~deps:()
          in
          (match result with
          | Error error -> expect error |. toBe input
          | Ok _ -> unreachable ());
          expect f |. not |. toBeCalled;
          Js.Promise.resolve ())
        |> asyncProperty1 (string ())
        |> asyncAssert
      in
      Js.Promise.resolve ());

  testAsync "catch" (fun () ->
      let open Promise.Monad_syntax in
      let () =
        (fun input ->
          match
            Error input
            |. catch (fun n -> Ok (n |> Js.Float.toString))
            |. run ~deps:()
          with
          | Error _ -> unreachable ()
          | Ok value -> expect value |. toBe (input |> Js.Float.toString))
        |> property1 (double ())
        |> assert_
      in
      let () =
        (fun input ->
          let f = fn1 () in
          (match Ok input |. catch f |. run ~deps:() with
          | Error _ -> unreachable ()
          | Ok value -> expect value |. toBe input);
          expect f |. not |. toBeCalled)
        |> property1 (string ())
        |> assert_
      in
      let* () =
        (fun input ->
          let* result =
            Async (fun () -> Error input |. Js.Promise.resolve)
            |. catch (fun n -> Ok (n |. Js.Float.toString))
            |. runAsync ~deps:()
          in
          (match result with
          | Error _ -> unreachable ()
          | Ok value -> expect value |. toBe (input |. Js.Float.toString));
          Js.Promise.resolve ())
        |> asyncProperty1 (double ())
        |> asyncAssert
      in
      let* () =
        (fun input ->
          let f = fn1 () in
          let* result =
            Async (fun () -> Ok input |> Js.Promise.resolve)
            |. catch f |. runAsync ~deps:()
          in
          (match result with
          | Error _ -> unreachable ()
          | Ok value -> expect value |. toBe input);
          expect f |. not |. toBeCalled;
          Js.Promise.resolve ())
        |> asyncProperty1 (string ())
        |> asyncAssert
      in
      Js.Promise.resolve ());

  test "ask" (fun () ->
      (fun thing ->
        match ask () |. run ~deps:thing with
        | Error _ -> unreachable ()
        | Ok env -> expect env |. toBe thing)
      |> property1 (anything ())
      |> assert_);

  test "provide" (fun () ->
      (fun thing ->
        match provide (ask ()) ~deps:thing |. run ~deps:() with
        | Error _ -> unreachable ()
        | Ok env -> expect env |. toBe thing)
      |> property1 (anything ())
      |> assert_)

File: src/fx/fx.spec.ts
import * as fc from 'fast-check';
import { expect, test, vi } from 'vitest';

import * as Fx from '../fx/fx';

test('Ok', () => {
  fc.assert(
    fc.property(fc.anything(), input => {
      const value = Fx.run(Fx.Ok(input), undefined);
      if (value.TAG === 0) expect(value._0).toBe(input);
      else expect.unreachable();
    }),
  );
});

test('Error', () => {
  fc.assert(
    fc.property(fc.anything(), input => {
      const error = Fx.run(Fx.Err(input), undefined);
      if (error.TAG === 0) expect.unreachable();
      expect(error._0).toBe(input);
    }),
  );
});

test('async', () =>
  fc
    .assert(
      fc.asyncProperty(fc.anything(), input =>
        Fx.runAsync(
          Fx.Async(async n => Fx.Ok(n)),
          input,
        ).then(result => {
          if (result.TAG === 0) expect(result._0).toBe(input);
          else expect.unreachable();
        }),
      ),
    )
    .then(() =>
      fc.assert(
        fc.asyncProperty(fc.anything(), input =>
          Fx.runAsync(
            Fx.Async(async n => Fx.Err(n)),
            input,
          ).then(result => {
            if (result.TAG === 0) expect.unreachable();
            else expect(result._0).toBe(input);
          }),
        ),
      ),
    ));

test('bind', () => {
  fc.assert(
    fc.property(fc.double(), input => {
      const value = Fx.run(
        Fx.bind(Fx.Ok(input), n => Fx.Ok(n.toString())),
        undefined,
      );

      if (value.TAG === 0) expect(value._0).toBe(String(input));
      else expect.unreachable();
    }),
  );

  fc.assert(
    fc.property(fc.string(), input => {
      const f = vi.fn();
      const error = Fx.run(Fx.bind(Fx.Err(input), f), undefined);
      if (error.TAG === 0) expect.unreachable();
      else expect(error._0).toBe(input);

      expect(f).not.toHaveBeenCalled();
    }),
  );

  return fc
    .assert(
      fc.asyncProperty(fc.double(), input => {
        return Fx.runAsync(
          Fx.bind(
            Fx.Async(async () => Fx.Ok(input)),
            n => Fx.Ok(n.toString()),
          ),
          undefined,
        ).then(result => {
          if (result.TAG === 0) expect(result._0).toBe(String(input));
          else expect.unreachable();
        });
      }),
    )
    .then(() =>
      fc.assert(
        fc.asyncProperty(fc.string(), input => {
          const f = vi.fn();

          return Fx.runAsync(
            Fx.bind(
              Fx.Async(async () => Fx.Err(input)),
              f,
            ),
            undefined,
          ).then(result => {
            if (result.TAG === 0) expect.unreachable();
            else expect(result._0).toBe(input);
            expect(f).not.toHaveBeenCalled();
          });
        }),
      ),
    );
});

test('map', () => {
  fc.assert(
    fc.property(fc.double(), input => {
      const value = Fx.run(
        Fx.map(Fx.Ok(input), n => String(n)),
        undefined,
      );

      if (value.TAG === 0) expect(value._0).toBe(input.toString());
      else expect.unreachable();
    }),
  );

  fc.assert(
    fc.property(fc.string(), input => {
      const f = vi.fn();
      const result = Fx.run(Fx.bind(Fx.Err(input), f), undefined);
      if (result.TAG === 0) expect.unreachable();
      else expect(result._0).toBe(input);

      expect(f).not.toHaveBeenCalled();
    }),
  );

  return fc
    .assert(
      fc.asyncProperty(fc.double(), input => {
        return Fx.runAsync(
          Fx.bind(Fx.Ok(input), n => Fx.Ok(n.toString())),
          undefined,
        ).then(result => {
          if (result.TAG === 0) expect(result._0).toBe(input.toString());
          else expect.unreachable();
        });
      }),
    )
    .then(() =>
      fc.assert(
        fc.asyncProperty(fc.string(), input => {
          const f = vi.fn();

          return Fx.runAsync(
            Fx.bind(
              Fx.Async(async () => Fx.Err(input)),
              f,
            ),
            undefined,
          ).then(result => {
            if (result.TAG === 0) expect.unreachable();
            else expect(result._0).toBe(input);

            expect(f).not.toHaveBeenCalled();
          });
        }),
      ),
    );
});

test('catch', () => {
  fc.assert(
    fc.property(fc.double(), input => {
      const value = Fx.run(
        Fx.catch_(Fx.Err(input), n => Fx.Ok(n.toString())),
        undefined,
      );

      if (value.TAG === 0) expect(value._0).toBe(input.toString());
      else expect.unreachable();
    }),
  );

  fc.assert(
    fc.property(fc.string(), input => {
      const f = vi.fn();

      const value = Fx.run(Fx.catch_(Fx.Ok(input), f), undefined);

      if (value.TAG === 0) expect(value._0).toBe(input);
      else expect.unreachable();

      expect(f).not.toHaveBeenCalled();
    }),
  );

  return fc
    .assert(
      fc.asyncProperty(fc.double(), input => {
        return Fx.runAsync(
          Fx.catch_(
            Fx.Async(async () => Fx.Err(input)),
            n => Fx.Ok(n.toString()),
          ),
          undefined,
        ).then(result => {
          if (result.TAG === 0) expect(result._0).toBe(input.toString());
          else expect.unreachable();
        });
      }),
    )
    .then(() =>
      fc.assert(
        fc.asyncProperty(fc.string(), input => {
          const f = vi.fn();

          return Fx.runAsync(
            Fx.catch_(
              Fx.Async(async () => Fx.Ok(input)),
              f,
            ),
            undefined,
          ).then(result => {
            if (result.TAG === 0) expect(result._0).toBe(input);
            else expect.unreachable();

            expect(f).not.toHaveBeenCalled();
          });
        }),
      ),
    );
});

test('ask', () => {
  fc.assert(
    fc.property(fc.anything(), thing => {
      const env = Fx.run(Fx.ask(), thing);
      if (env.TAG === 0) expect(env._0).toBe(thing);
      else expect.unreachable();
    }),
  );
});

test('provide', () => {
  fc.assert(
    fc.property(fc.anything(), thing => {
      const env = Fx.run(Fx.provide(Fx.ask(), thing), undefined);
      if (env.TAG === 0) expect(env._0).toBe(thing);
      else expect.unreachable();
    }),
  );
});
